swifttool/manager.py

```diff
@@ -0,0 +1,288 @@
+
+    def _write_status(self, percent_done):
+        with open(self.current_status, 'w') as status:
+            status.write(str(percent_done))
+
+    def _update_cluster_nodes(self):
+        try:
+            updated_cfg = copy.deepcopy(self.cluster_config)
+            # Update the cluster config to what
+            # we desirei
+            changed = False
+            for builder in RING_TYPES:
+                if builder not in self.cluster_changes:
+                    continue
+                LOG.debug("Updating %s builder.", builder)
+                for builder_changes in self.cluster_changes[builder]:
+                    for zone, devices in builder_changes.items():
+                        # Get the current configuration from rings
+                        ring_devs = self._get_ring_devices(builder, zone)
+                        ip = self._get_zone_ip(devices)
+                        updated_cfg['zones'].setdefault(zone, {})
+                        updated_cfg['zones'][zone].setdefault(ip, {})
+                        updated_cfg['zones'][zone][ip].setdefault('disks', {})
+                        newdevs = []
+                        for dev in devices:
+                            devname = dev['device']
+                            if devname not in ring_devs:
+                                initial_weight = 0
+                            else:
+                                initial_weight = ring_devs[devname]
+                            if initial_weight == dev['target']:
+                                continue
+                            devinfo = {}
+                            if 'metadata' in dev:
+                                devinfo['metadata'] = dev['metadata']
+                            devinfo['blockdev'] = devname
+                            devinfo['weight'] = int(initial_weight +
+                                                    dev['step'])
+                            if abs(devinfo['weight'] - dev['target']) < 50:
+                                devinfo['weight'] = dev['target']
+                            LOG.debug("Setting the %s device capacity to %s",
+                                      devname, str(devinfo['weight']))
+                            newdevs.append(devinfo)
+                            changed = True
+                        updated_cfg['zones'][zone][ip]['disks'][builder] = \
+                            newdevs
+            if changed:
+                ringsdef_helper(updated_cfg, self.metadata, self.outdir)
+            return changed, updated_cfg
+        except Exception as e:
+            print "Error updating cluster configuration", e
+            sys.exit(-1)
+
+    @staticmethod
+    def _get_zone_ip(devs):
+        for dev in devs:
+            if 'ip' in dev:
+                return dev['ip']
+
+    def _calculate_cluster_changes(self):
+        try:
+            self.cluster_config = yaml.load(open(self.def_file, 'r'))
```

Comment:
i'd use the open context manager again rather than leaving this stream open, which i think it does.